// sculptlyWeb/index.js
///
/// The web module handles all requests from a browser for Sculptly.
///
/// Exported Functions
///   - `start` Turns on the web server.
///

// Load required modules
var m = {
    // Node modules
    http : require('http'),
    url  : require('url'),
    path : require('path'),
    fs   : require('fs'),

    // Sculptly modules
    api  : require('./api')
};

// Set up our globals
var port = process.env.C9_PORT;
var host = '0.0.0.0';
var webServer = null;

// --- Exported Functions --- //

exports.start = function(){
    console.log( '*** Starting Server ***' );

    webServer = m.http.createServer();
    webServer.listen( port, host );
    webServer.on( 'request', onRequest );
    webServer.on( 'connection', onConnection );
    webServer.on( 'close', onClose );
    webServer.on( 'upgrade', onUpgrade );

    console.log( '*** Server Started ***' );
};

// --- Internal Functions --- //

/// Handles the HTTP request event.
///
/// This decides if the request needs to be handled directly by this module (such
/// as file requests) or passed off to a different module to handle (such as API
/// calls).
///
/// @param {ServerRequest}  req The request.
/// @param {ServerResponse} res Object to write the response to.
function onRequest( req, res ){
    // Is this request for the api or for us?
    var filename = m.url.parse( req.url ).pathname;
    if( filename.match( /^\/api$/ ) )
        m.api.emit( 'request', req, res );
    else
        handlePageRequest( res, filename );
}

/// Handles a file request.
///
/// Requests for JS, CSS, and HTML files all pass through this function.
///
/// @param {ServerResponse} res      The response object to send the file
///                         contents through.
/// @param {String}         filename The name of the file to retrieve.
function handlePageRequest( res, filename )
{
    if( !filename.length || filename == '/' )
        filename = 'index.html';
    var fullPath = m.path.join( __dirname, 'client', filename );
    m.path.exists( fullPath, function( exists ){
        if( !exists )
            respond404( res );
        else
            respondFile( res, fullPath );
    });
}

/// Responds with a 404, page-not-found.
///
/// @param {ServerResponse} res The response object to send a 404 to.
function respond404( res ){
    res.writeHead( 404, { 'Content-Type' : 'text/plain' } );
    res.end( 'File not found' );
}

/// Responds with a specific file's contents.
///
/// @param {ServerResponse} res      The response object to send the file
///                         contents through.
/// @param {String}         filePath The name of the file to retrieve.
function respondFile( res, filePath ){
    // Get the content type
    var contentType = 'text/html';
    if( filePath.match( /\.js$/ ) )
        contentType = 'application/javascript';
    else if( filePath.match( /\.css$/ ) )
        contentType = 'text/css';

    // Write our header and pipe the file through
    res.writeHead( 200, { 'Content-Type' : contentType } );
    var stream = m.fs.createReadStream( filePath, { encoding : 'utf8' } );
    stream.pipe( res );
}

function onConnection( stream ){
    console.log( 'Connection opened' );
}

function onClose( stream ){
    console.log( 'Connection closed' );
}

function onUpgrade( request, socket, head ){
    console.log( 'Connection upgraded' );
}
